frontend/test/resolution/testTypePropertyPrimitives.cpp

```diff
@@ -0,0 +1,580 @@
+    TRUE,
+    FALSE,
+    STRING,
+    ERROR 
+  };
+
+  struct PrimitiveCalls {
+    std::vector<const char*> actuals;
+    Output expected;
+    std::string str;
+
+    PrimitiveCalls(std::vector<const char*> actuals, Output expected,
+                   std::string str="")
+      : actuals(std::move(actuals)),
+        expected(expected),
+        str(std::move(str)) {
+      if (expected != STRING) assert(this->str.empty());
+    }
+  };
+
+  std::string testName;
+  bool isChplHomeRequired = false;
+  std::string prelude;
+  chpl::uast::primtags::PrimitiveTag primitive;
+  std::vector<PrimitiveCalls> calls;
+};
+
+static bool
+isParamStringMatch(chpl::types::QualifiedType qt, std::string str,
+                   std::string* out) {
+  if (qt.kind() == QualifiedType::PARAM) {
+    if (auto t = qt.type()) {
+      if (auto p = qt.param()) {
+        if (auto sp = p->toStringParam()) {
+          if (out) *out = sp->value().c_str();
+          return sp->value() == str;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+static void
+assertParamStringMatch(chpl::types::QualifiedType qt, std::string str,
+                       std::string* out) {
+  bool match = isParamStringMatch(qt, str, out);
+  if (!match) {
+    std::cout << "Expected " << str << ", but got -> " << *out << std::endl;
+    std::abort();
+  }
+}
+
+static void testPrimitive(Test tpg) {
+  Context::Configuration config;
+  if (tpg.isChplHomeRequired) {
+    if (const char* chplHomeEnv = getenv("CHPL_HOME")) {
+      config.chplHome = chplHomeEnv;
+    } else {
+      std::cout << "CHPL_HOME must be set!" << std::endl;
+      exit(1);
+    }
+  }
+  Context ctx(config);
+  Context* context = &ctx;
+  ErrorGuard guard(context);
+
+  std::stringstream ps;
+  int counter = 0;
+  int expectedErrorCount = 0;
+  std::vector<std::string> variables;
+  std::vector<Test::Output> expected;
+  const auto tagStr = chpl::uast::primtags::primTagToName(tpg.primitive);
+
+  ps << tpg.prelude << std::endl;
+
+  for (auto& call : tpg.calls) {
+    if (call.expected == Test::ERROR) expectedErrorCount += 1;
+
+    auto var = std::string("x_") + std::to_string(counter++);
+    variables.push_back(var);
+    expected.push_back(call.expected);
+
+    ps << "param " << var << " = " << "__primitive(\"";
+    ps << tagStr << "\", ";
+
+    for (int i = 0; i < call.actuals.size(); i++) {
+      ps << call.actuals[i];
+      bool last = (i+1) == call.actuals.size();
+      if (!last) ps << ", ";
+    }
+    ps << ");" << std::endl;
+  }
+
+  auto text = ps.str();
+  std::cout << "--- program " << tpg.testName << " (" << tagStr << ") ---";
+  std::cout << std::endl << text << "---" << std::endl;
+  auto varTypes = resolveTypesOfVariables(context, text, variables);
+
+  for (size_t i = 0; i < varTypes.size(); i++) {
+    auto& call = tpg.calls[i];
+    auto& var = variables[i];
+    auto& type = varTypes.at(var);
+    std::cout << "Checking " << var << " -> ";
+
+    switch (expected[i]) {
```

Comment:
Do you need `expected[i]` and `expected` in general if you can just get it from `tpg.calls[i].expected`?